In [1]:
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

# Data cleaning

## Loading the CSV file

In [42]:
path = ""
file_name = "OpenFood.csv"
data_path = path+file_name
df = pd.read_csv(data_path, sep = '\t')

/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (0,3,5,24,25,26,28,36,41,49) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
df.to_pickle("openFood.pkl")

In [4]:
df = pd.read_pickle("openFood.pkl")

## DataFrame Exploration

In [5]:
df.head()

code                                                url   creator  \
0  0000000000017  http://world-en.openfoodfacts.org/product/0000...   kiliweb   
1  0000000000031  http://world-en.openfoodfacts.org/product/0000...  isagoofy   
2  0000000000123  http://world-en.openfoodfacts.org/product/0000...   kiliweb   
3  0000000000178  http://world-en.openfoodfacts.org/product/0000...   kiliweb   
4  0000000000291  http://world-en.openfoodfacts.org/product/0000...   kiliweb   

    created_t      created_datetime last_modified_t last_modified_datetime  \
0  1529059080  2018-06-15T10:38:00Z      1529059204   2018-06-15T10:40:04Z   
1  1539464774  2018-10-13T21:06:14Z      1539464817   2018-10-13T21:06:57Z   
2  1535737982  2018-08-31T17:53:02Z      1535737986   2018-08-31T17:53:06Z   
3  1542456332  2018-11-17T12:05:32Z      1542456333   2018-11-17T12:05:33Z   
4  1534239669  2018-08-14T09:41:09Z      1534239732   2018-08-14T09:42:12Z   

            product_name generic_name quantity      ...        \
0       Vitória crackers          NaN      NaN      ...         
1                  Cacao          NaN    130 g      ...         
2  Sauce Sweety chili 0%          NaN      NaN      ...         
3              Mini coco          NaN      NaN      ...         
4              Mendiants          NaN      NaN      ...         

  carbon-footprint_100g nutrition-score-fr_100g nutrition-score-uk_100g  \
0                   NaN                     NaN                     NaN   
1                   NaN                     NaN                     NaN   
2                   NaN                     NaN                     NaN   
3                   NaN                     NaN                     NaN   
4                   NaN                     NaN                     NaN   

  glycemic-index_100g water-hardness_100g choline_100g phylloquinone_100g  \
0                 NaN                 NaN          NaN                NaN   
1                 NaN                 NaN          NaN                NaN   
2                 NaN                 NaN          NaN                NaN   
3                 NaN                 NaN          NaN                NaN   
4                 NaN                 NaN          NaN                NaN   

  beta-glucan_100g inositol_100g carnitine_100g  
0              NaN           NaN            NaN  
1              NaN           NaN            NaN  
2              NaN           NaN            NaN  
3              NaN           NaN            NaN  
4              NaN           NaN            NaN  

[5 rows x 173 columns]

### DataFrame properties

Are there some duplicated rows?

In [6]:
np.sum(df.duplicated().values == True)

26

So the duplicated values are deleted:

In [7]:
df.drop_duplicates(inplace = True)

We can also drop the column that only conain NaN values:

In [8]:
df.dropna(axis = 1, how='all', inplace = True)

In [9]:
df.shape

(699098, 169)

So now there are 699098 rows and 169 columns

Now can the code columnm be set as the index. In order to be so it must be unique:

In [10]:
df.code.is_unique

False

The codes are not unique so let's check what those values are:

In [11]:
duplicate_codes_indices = np.where(df.duplicated('code', keep=False).values)[0]
dups = df.iloc[duplicate_codes_indices].copy()
dups.code = dups.code.astype(float)
dups.sort_values(by = 'code', inplace = True)
dups

code                                                url  \
66475   5.000000e+01  http://world-en.openfoodfacts.org/product/0050...   
535902  5.000000e+01       http://world-en.openfoodfacts.org/product/50   
596588  7.340000e+02      http://world-en.openfoodfacts.org/product/734   
167953  7.340000e+02  http://world-en.openfoodfacts.org/product/0734...   
10873   1.160220e+05  http://world-en.openfoodfacts.org/product/0011...   
209616  1.160220e+05  http://world-en.openfoodfacts.org/product/1160...   
148839  6.356460e+05  http://world-en.openfoodfacts.org/product/0635...   
75132   6.356460e+05  http://world-en.openfoodfacts.org/product/0063...   
75134   6.356840e+05  http://world-en.openfoodfacts.org/product/0063...   
148845  6.356840e+05  http://world-en.openfoodfacts.org/product/0635...   
165192  7.228100e+05  http://world-en.openfoodfacts.org/product/0722...   
89530   7.228100e+05  http://world-en.openfoodfacts.org/product/0072...   
112326  8.379960e+05  http://world-en.openfoodfacts.org/product/0083...   
188264  8.379960e+05  http://world-en.openfoodfacts.org/product/0837...   
205342  8.887070e+05  http://world-en.openfoodfacts.org/product/0888...   
117435  8.887070e+05  http://world-en.openfoodfacts.org/product/0088...   
207921  8.984780e+05  http://world-en.openfoodfacts.org/product/0898...   
118000  8.984780e+05  http://world-en.openfoodfacts.org/product/0089...   
141754  3.318233e+06  http://world-en.openfoodfacts.org/product/0331...   
362674  3.318233e+06  http://world-en.openfoodfacts.org/product/3318...   
520415  4.159932e+06  http://world-en.openfoodfacts.org/product/4159...   
142651  4.159932e+06  http://world-en.openfoodfacts.org/product/0415...   
209301  1.111082e+10  http://world-en.openfoodfacts.org/product/1111...   
4306    1.111082e+10  http://world-en.openfoodfacts.org/product/0011...   
12141   1.356200e+10  http://world-en.openfoodfacts.org/product/0013...   
210444  1.356200e+10  http://world-en.openfoodfacts.org/product/1356...   
210445  1.356200e+10  http://world-en.openfoodfacts.org/product/1356...   
12142   1.356200e+10  http://world-en.openfoodfacts.org/product/0013...   
12535   1.410007e+10  http://world-en.openfoodfacts.org/product/0014...   
210625  1.410007e+10  http://world-en.openfoodfacts.org/product/1410...   
...              ...                                                ...   
265815           NaN                                                NaN   
265826           NaN                                                NaN   
265850           NaN                                                NaN   
265863           NaN                                                NaN   
265871           NaN                                                NaN   
265873           NaN                                                NaN   
265879           NaN                                                NaN   
265926           NaN                                                NaN   
265946           NaN                                                NaN   
265956           NaN                                                NaN   
265958           NaN                                                NaN   
265962           NaN                                                NaN   
265964           NaN                                                NaN   
265975           NaN                                                NaN   
265977           NaN                                                NaN   
265984           NaN                                                NaN   
265987           NaN                                                NaN   
266023           NaN                                                NaN   
266065           NaN                                                NaN   
266068           NaN                                                NaN   
266098           NaN                                                NaN   
266100           NaN                     

As it can be seen there are two types of duplicates:
Some codes appear twice and the some code are NaN. As it can be observed when code are duplicated they are often the same and sometimes one row contain more information than the other. So we could just keep the rows that contain the most information. And for the rows with NaN as code, the values of the columns are not matching the column so there must be a shift of values. As a lot of values are missing, it would be complex to recover the true data and there are only tens of code with NaNs, we can delete them:

In [21]:
dups = dups[dups != float('nan')]

dups['nb_of_features'] = np.sum(dups.notnull(), axis=1)
dups.reset_index(inplace = True)
indices_to_remove = dups.sort_values(['code','nb_of_features'],ascending = True)\
                        .groupby('code',as_index=False).first()['index'].values
df.dropna(subset=['code'],inplace=True)
df.drop(indices_to_remove,inplace = True)

In [22]:
df.code.is_unique

True

In [25]:
df.set_index('code',inplace = True)
df.head()

url   creator  \
code                                                                         
0000000000017  http://world-en.openfoodfacts.org/product/0000...   kiliweb   
0000000000031  http://world-en.openfoodfacts.org/product/0000...  isagoofy   
0000000000123  http://world-en.openfoodfacts.org/product/0000...   kiliweb   
0000000000291  http://world-en.openfoodfacts.org/product/0000...   kiliweb   
0000000000949  http://world-en.openfoodfacts.org/product/0000...   kiliweb   

                created_t      created_datetime last_modified_t  \
code                                                              
0000000000017  1529059080  2018-06-15T10:38:00Z      1529059204   
0000000000031  1539464774  2018-10-13T21:06:14Z      1539464817   
0000000000123  1535737982  2018-08-31T17:53:02Z      1535737986   
0000000000291  1534239669  2018-08-14T09:41:09Z      1534239732   
0000000000949  1523440813  2018-04-11T10:00:13Z      1523440823   

              last_modified_datetime               product_name generic_name  \
code                                                                           
0000000000017   2018-06-15T10:40:04Z           Vitória crackers          NaN   
0000000000031   2018-10-13T21:06:57Z                      Cacao          NaN   
0000000000123   2018-08-31T17:53:06Z      Sauce Sweety chili 0%          NaN   
0000000000291   2018-08-14T09:42:12Z                  Mendiants          NaN   
0000000000949   2018-04-11T10:00:23Z  Salade de carottes râpées          NaN   

              quantity packaging      ...       carbon-footprint_100g  \
code                                  ...                               
0000000000017      NaN       NaN      ...                         NaN   
0000000000031    130 g       NaN      ...                         NaN   
0000000000123      NaN       NaN      ...                         NaN   
0000000000291      NaN       NaN      ...                         NaN   
0000000000949      NaN       NaN      ...                         NaN   

              nutrition-score-fr_100g nutrition-score-uk_100g  \
code                                                            
0000000000017                     NaN                     NaN   
0000000000031                     NaN                     NaN   
0000000000123                     NaN                     NaN   
0000000000291                     NaN                     NaN   
0000000000949                     NaN                     NaN   

              glycemic-index_100g water-hardness_100g choline_100g  \
code                                                                 
0000000000017                 NaN                 NaN          NaN   
0000000000031                 NaN                 NaN          NaN   
0000000000123                 NaN                 NaN          NaN   
0000000000291                 NaN                 NaN          NaN   
0000000000949                 NaN                 NaN          NaN   

              phylloquinone_100g beta-glucan_100g inositol_100g carnitine_100g  
code                                                                            
0000000000017                NaN              NaN           NaN            NaN  
0000000000031                NaN              NaN           NaN            NaN  
0000000000123                NaN              NaN           NaN            NaN  
0000000000291                NaN              NaN           NaN            NaN  
0000000000949                NaN              NaN           NaN            NaN  

[5 rows x 168 columns]

In [18]:
pd.set_option('display.max_columns', 500)

### Cleaning columns

In order to answer our research questions, we do not need all the columns such as *creator* or *created_datetime* among others. We decided to drop all the columns that are not needed

In [60]:
columns_to_drop = ['url',
                  'creator',
                  'created_t',
                  'created_datetime',
                  'last_modified_t',
                  'last_modified_datetime',
                  'generic_name',
                  'quantity',
                  'countries_tags',
                  'brands_tags',
                  'origins',
                  'countries',
                  'origins_tags',
                  'emb_codes',
                  'emb_codes_tags',
                  'first_packaging_code_geo',
                  'cities',
                  'cities_tags',
                  'purchase_places',
                  'stores',
                  'traces',
                  'traces_tags',
                  'serving_size',
                  'no_nutriments',
                  'additives',
                  'additives_tags',
                  'image_url',
                  'allergens', 
                  'image_small_url',
                  'ph_100g',
                  ]

In [64]:
df.drop(columns_to_drop, axis=1).columns

Index(['code', 'product_name', 'packaging', 'packaging_tags', 'brands',
       'categories', 'categories_tags', 'categories_en',
       'manufacturing_places', 'manufacturing_places_tags',
       ...
       'carbon-footprint_100g', 'nutrition-score-fr_100g',
       'nutrition-score-uk_100g', 'glycemic-index_100g', 'water-hardness_100g',
       'choline_100g', 'phylloquinone_100g', 'beta-glucan_100g',
       'inositol_100g', 'carnitine_100g'],
      dtype='object', length=143)

code   creator   created_t      created_datetime last_modified_t  \
0  0000000000017   kiliweb  1529059080  2018-06-15T10:38:00Z      1529059204   
1  0000000000031  isagoofy  1539464774  2018-10-13T21:06:14Z      1539464817   
2  0000000000123   kiliweb  1535737982  2018-08-31T17:53:02Z      1535737986   
3  0000000000178   kiliweb  1542456332  2018-11-17T12:05:32Z      1542456333   
4  0000000000291   kiliweb  1534239669  2018-08-14T09:41:09Z      1534239732   

  last_modified_datetime           product_name generic_name quantity  \
0   2018-06-15T10:40:04Z       Vitória crackers          NaN      NaN   
1   2018-10-13T21:06:57Z                  Cacao          NaN    130 g   
2   2018-08-31T17:53:06Z  Sauce Sweety chili 0%          NaN      NaN   
3   2018-11-17T12:05:33Z              Mini coco          NaN      NaN   
4   2018-08-14T09:42:12Z              Mendiants          NaN      NaN   

  packaging packaging_tags brands brands_tags categories categories_tags  \
0       NaN            NaN    NaN         NaN        NaN             NaN   
1       NaN            NaN    NaN         NaN        NaN             NaN   
2       NaN            NaN    NaN         NaN        NaN             NaN   
3       NaN            NaN    NaN         NaN        NaN             NaN   
4       NaN            NaN    NaN         NaN        NaN             NaN   

  categories_en origins origins_tags manufacturing_places  \
0           NaN     NaN          NaN                  NaN   
1           NaN     NaN          NaN                  NaN   
2           NaN     NaN          NaN                  NaN   
3           NaN     NaN          NaN                  NaN   
4           NaN     NaN          NaN                  NaN   

  manufacturing_places_tags labels labels_tags labels_en emb_codes  \
0                       NaN    NaN         NaN       NaN       NaN   
1                       NaN    NaN         NaN       NaN       NaN   
2                       NaN    NaN         NaN       NaN       NaN   
3                       NaN    NaN         NaN       NaN       NaN   
4                       NaN    NaN         NaN       NaN       NaN   

  emb_codes_tags first_packaging_code_geo cities_tags purchase_places stores  \
0            NaN                      NaN         NaN             NaN    NaN   
1            NaN                      NaN         NaN             NaN    NaN   
2            NaN                      NaN         NaN             NaN    NaN   
3            NaN                      NaN         NaN             NaN    NaN   
4            NaN                      NaN         NaN             NaN    NaN   

   countries countries_tags countries_en ingredients_text allergens  \
0  en:france      en:france       France              NaN       NaN   
1      en:FR      en:france       France              NaN       NaN   
2  en:france      en:france       France              NaN       NaN   
3  en:france      en:france       France              NaN       NaN   
4  en:france      en:france       France              NaN       NaN   

  allergens_en traces traces_tags traces_en serving_size serving_quantity  \
0          NaN    NaN         NaN       NaN          NaN                0   
1          NaN    NaN         NaN       NaN          NaN                0   
2          NaN    NaN         NaN       NaN          NaN                0   
3          NaN    NaN         NaN       NaN          NaN                0   
4          NaN    NaN         NaN       NaN          NaN                0   

   no_nutriments  additives_n additives additives_tags additives_en  \
0            NaN          NaN       NaN            NaN          NaN   
1            NaN          NaN       NaN            NaN          NaN   
2            NaN          NaN       NaN            NaN          NaN   
3            NaN          NaN       NaN            NaN          NaN   
4            NaN          NaN       NaN            NaN          NaN   

   ingredients_from_palm_oil_n ingredients_from_p